File: Src2/Icm.Core/Functions/TotalOrder/ITotalOrderExtensions.vb
Imports System.Runtime.CompilerServices

Namespace Icm
    Public Module ITotalOrderExtensions

        <Extension()>
        Public Function LstIfNull(Of T As {Structure, IComparable(Of T)})(ByVal totalOrder As ITotalOrder(Of T), ByVal obj As T?) As T
            If obj.HasValue Then
                Return obj.Value
            Else
                Return totalOrder.Least
            End If
        End Function

        <Extension()>
        Public Function GstIfNull(Of T As {Structure, IComparable(Of T)})(ByVal totalOrder As ITotalOrder(Of T), ByVal obj As T?) As T
            If obj.HasValue Then
                Return obj.Value
            Else
                Return totalOrder.Greatest
            End If
        End Function

        <Extension()>
        Public Function LstIfNull(Of T As {IComparable(Of T)})(ByVal totalOrder As ITotalOrder(Of T), ByVal obj As Nullable2(Of T)) As T
            If obj.HasValue Then
                Return obj.Value
            Else
                Return totalOrder.Least
            End If
        End Function

        <Extension()>
        Public Function GstIfNull(Of T As {IComparable(Of T)})(ByVal totalOrder As ITotalOrder(Of T), ByVal obj As Nullable2(Of T)) As T
            If obj.HasValue Then
                Return obj.Value
            Else
                Return totalOrder.Greatest
            End If
        End Function

    End Module
End Namespace

File: Src/Icm.Core/Functions/TotalOrder/ITotalOrderExtensions.cs
using System;

namespace Icm
{
    public static class TotalOrderExtensions
    {
        public static T LstIfNull<T>(this ITotalOrder<T> totalOrder, T? obj) where T : struct, IComparable<T>
        {
            return obj ?? totalOrder.Least();
        }

        public static T GstIfNull<T>(this ITotalOrder<T> totalOrder, T? obj) where T : struct, IComparable<T>
        {
            return obj ?? totalOrder.Greatest();
        }

        public static T LstIfNull<T>(this ITotalOrder<T> totalOrder, Nullable2<T> obj) where T : IComparable<T>
        {
            return obj.HasValue 
                ? obj.Value
                : totalOrder.Least();
        }

        public static T GstIfNull<T>(this ITotalOrder<T> totalOrder, Nullable2<T> obj) where T : IComparable<T>
        {
            return obj.HasValue
                ? obj.Value
                : totalOrder.Greatest();
        }

    }
}